componentLibraries/extensionLibrary/Connectivity/FMIWrapper.hpp

```diff
@@ -0,0 +1,355 @@
+{
+    hopsan::Component* pComponent = (hopsan::Component*)pComponentEnvironment;
+    if (pComponent == nullptr) {
+        return;
+    }
+
+    char buffer[512];
+    va_list args;
+    va_start(args, message);
+    vsnprintf(buffer, 512, message, args);
+    va_end(args);
+
+    switch (status) {
+    // Typically info messages are not something we want to see in Hopsan, so show them as debug type
+    case fmi2_status_ok:
+        pComponent->addDebugMessage(buffer);
+        break;
+    case fmi2_status_warning:
+        pComponent->addWarningMessage(buffer);
+        break;
+    case fmi2_status_error:
+        pComponent->addErrorMessage(buffer);
+        break;
+    case fmi2_status_fatal:
+        pComponent->addFatalMessage(buffer);
+        break;
+    default:
+        // Discard
+        break;
+    }
+}
+
+
+
+class FMIWrapper : public ComponentSignal
+{
+private:
+    HString mFmuPath, mLastFmuPath, mTmpPath;
+    std::map<fmi2_value_reference_t,double*> mOutputs;
+    std::map<fmi2_value_reference_t,double*> mInputs;
+    std::map<fmi2_value_reference_t,double> mParameters;
+    std::vector<Port*> mPorts;
+
+    jm_callbacks callbacks;
+    jm_status_enu_t status;
+    fmi_import_context_t* context;
+    fmi_version_enu_t version;
+    fmi2_callback_functions_t fmiCallbackFunctions;
+    fmi2_status_t fmistatus;
+    fmi2_import_t* fmu;
+
+
+public:
+    static Component *Creator()
+    {
+        return new FMIWrapper();
+    }
+
+    void configure()
+    {
+        addConstant("path", "Path to functional mockup unit (FMU)", "", mFmuPath);
+    }
+
+    void reconfigure()
+    {
+        if(mFmuPath == mLastFmuPath) {
+            return; //Path did not change, do nothing
+        }
+        mLastFmuPath = mFmuPath;
+
+        deconfigure(); //Make sure to unload FMU and free memory before loading a new one
+
+        for(const auto *port : mPorts) {
+            removePort(port->getName());
+        }
+        std::vector<HString> parameters;
+        this->getParameterNames(parameters);
+        for(const auto &parameter : parameters) {
+            if(parameter != "path") {
+                this->unRegisterParameter(parameter);
+            }
+        }
+        mPorts.clear();
+        mOutputs.clear();
+        mInputs.clear();
+        mParameters.clear();
+
+        callbacks.malloc = malloc;
+        callbacks.calloc = calloc;
+        callbacks.realloc = realloc;
+        callbacks.free = free;
+        callbacks.logger = jmLogger;
+        callbacks.log_level = jm_log_level_debug;
+        callbacks.context = static_cast<jm_voidp>(this);
+
+        context = fmi_import_allocate_context(&callbacks);
+
+        addInfoMessage("Loading FMU from "+mFmuPath+"...");
+
+        //Generate random temporary path
+        mTmpPath = std::filesystem::temp_directory_path().c_str();
+        std::random_device rd;
+        std::default_random_engine gen(rd());
```

Comment:
This may also be a problem, is it available pre c++11 (I did not check)?
HopsanCore "must" be compilable with C++03 or the feature in this case must be optional, maybe could use old-style c random generator if pre c++11.

Reply:
Unfortunately not, I think it is actually C++14. We can use `rand()` instead. `rand() % 1000000` should generate a random number between 0 and 999999, which is what we need for a six-digit random number in the temp directory. I believe this is sufficient to avoid name collisions.